ddclient.in

```diff
@@ -7940,6 +7957,447 @@ sub nic_infomaniak_update {
+        my $value = $finalized_headers{$header_key};
+        $value =~ s/^\s+|\s+$//g;
+        debug("Key: ".$header_key." | Value: ".$value."\n");
+        $header_string = $header_string.(lc $header_key).':'.($value)."\n";
+        $signed_headers = $signed_headers.(lc $header_key);
+
+        if ($header_key ne $finalized_header_keys[$#finalized_header_keys]) {
+            $signed_headers = $signed_headers.';';
+        }
+
+        $formatted_finalized_headers{lc $header_key} = $value;        
+    }
+    debug("\n");
+
+    my $canonical_uri = %$url{"path"};
+    my $canonical_query_string = get_encoded_query_string($url);
+    debug("\nPayload: ".$payload_to_hash."\n");
+    my $canonical_request = "$method\n$canonical_uri\n$canonical_query_string\n$header_string\n$signed_headers\n$payload";
+
+    debug("Canonical Request\n-------------\n$canonical_request\n-------------\n\n");
+
+    my %result = (
+        hash => lc sha256_hex($canonical_request),
+        signed_headers => $signed_headers,
+        finalized_headers => \%formatted_finalized_headers
+    );
+
+    return \%result;
+}
+
+my $ALGORITHM = "AWS4-HMAC-SHA256";
+
+sub create_string_to_sign {
+    my (
+        $hash,
+        $service,
+        $region,
+        $date
+    ) = @_;
+
+    my $regionToUse = $region // $ENV{"AWS_REGION"} // "us-east-1";
+    
+    my $scope = to_short_date($date)."/$regionToUse/$service/aws4_request";
+    my %result = (
+        string => $ALGORITHM."\n".(to_iso_string($date))."\n$scope\n$hash",
+        scope => $scope
+    );
+
+    debug("String To Sign\n-------------\n".$result{"string"}."\n-------------\n");
+
+    return \%result;
+}
+    
+sub create_signature {
+    my (
+        $string_to_sign,
+        $secret_access_key,
+        $region,
+        $service,
+        $date
+    ) = @_;
+    debug("\n\nSecret: "."AWS4".$secret_access_key." | Short-Date: ".to_short_date($date)."\n\n");
+    debug("Region: ".$region." | Service: ".$service."\n\n");
+    my $k_date = hmac_sha256(to_short_date($date), "AWS4".$secret_access_key);
+    my $k_region = hmac_sha256($region, $k_date,);
+    my $k_service = hmac_sha256($service, $k_region);
+    my $k_signing = hmac_sha256("aws4_request", $k_service);
+    return lc hmac_sha256_hex($string_to_sign, $k_signing);
+}
+
+sub create_signed_request {
+    my %params = @_;
+    my $service = $params{"service"}; # In the future will likely parse this from the url
+    my $region = $params{'region'} // $ENV{"AWS_REGION"} // "us-east-1";
+    my $aws_secret_access_key = $params{'aws_secret_access_key'} // $ENV{"AWS_SECRET_ACCESS_KEY"};
+    my $aws_access_key_id = $params{'aws_access_key_id'} // $ENV{"AWS_ACCESS_KEY_ID"};
+    my $request_url = $params{'request_url'};
+    my $request_method = $params{'request_method'} // "GET";
+    my $request_payload = $params{'request_payload'} // "";
+    my $request_headers = $params{'request_headers'} // {};
+
+    my $url = parse_url($request_url);
```

Comment:
Since the only strictly invalid case for our purpose here is a missing protocol and host checked those are there. If not then an error is now thrown. Hopefully, that covers enough given the scope of this here.